transaction/LockManager.cpp

```diff
@@ -0,0 +1,217 @@
+          permitted_requests_.push(request);
+        }
+      }
+    }
+
+    // Resolve deadlocks.
+    killVictims();
+  }
+}
+
+bool LockManager::acquireLock(const transaction_id tid,
+                              const ResourceId &rid,
+                              const AccessMode access_mode) {
+  std::stack<std::pair<ResourceId, AccessMode>> stack;
+  ResourceId current_rid = rid;
+  AccessMode current_access_mode = access_mode;
+  stack.push(std::make_pair(current_rid, current_access_mode));
+
+  while (current_rid.hasParent()) {
+    current_rid = current_rid.getParentResourceId();
+    current_access_mode = (current_access_mode.isShareLock() ||
+                           current_access_mode.isIntentionShareLock())
+                              ? AccessMode(AccessModeType::kIsLock)
+                              : AccessMode(AccessModeType::kIxLock);
+
+    stack.push(std::make_pair(current_rid, current_access_mode));
+  }
+
+  lock_table_->latchExclusive();
+
+  while (!stack.empty()) {
+    const std::pair<ResourceId, AccessMode> pair_to_pick = stack.top();
+    const ResourceId rid_to_pick = pair_to_pick.first;
+    const AccessMode access_mode_to_pick = pair_to_pick.second;
+
+    if (!acquireLockInternal(tid, rid_to_pick, access_mode_to_pick)) {
+      lock_table_->unlatchExclusive();
+      return false;
+    }
+    stack.pop();
+  }
+  lock_table_->unlatchExclusive();
+  return true;
+}
+
+bool LockManager::releaseAllLocks(const transaction_id tid) {
+  const std::vector<ResourceId> resource_ids
+      = transaction_table_->getResourceIdList(tid);
+  const TransactionTableResult transaction_deleted
+      = transaction_table_->deleteTransaction(tid);
+  if (transaction_deleted
+          == TransactionTableResult::kTransactionDeleteError) {
+    LOG(FATAL)
+        << "In LockManager.releaseAllLocks: Transaction could not be deleted!";
+  }
+
+  lock_table_->latchExclusive();
+
+  for (const auto &resource_id : resource_ids) {
+    const LockTableResult lock_deleted = lock_table_->deleteLock(tid, resource_id);
+
+    LOG(INFO) << "Transaction "
+                 + std::to_string(tid)
+                 + " released lock:"
+                 + resource_id.toString()
+                 + "\n";
+
+    if (lock_deleted == LockTableResult::kDeleteError) {
+      LOG(FATAL) << "In LockManager.releaseAllLock lock could not be deleted "
+                    "from LockTable";
+    }
+  }
+
+  lock_table_->unlatchExclusive();
+  return true;
+}
+
+bool LockManager::acquireLockInternal(const transaction_id tid,
+                                      const ResourceId &rid,
+                                      const AccessMode access_mode) {
+  const LockTableResult lock_result = lock_table_->putLock(tid, rid, access_mode);
+  CHECK(lock_result != LockTableResult::kPutError)
+      << "Unexpected result in LockManager.acquireLockInternal";
+
+  if (lock_result == LockTableResult::kAlreadyInOwned) {
```

Comment:
I was wondering if you missed the brackets for cases `LockTableResult::kPlacedInOwned` and `LockTableResult::kPlacedInPending`. The following code passes on my Mac using `clang`:

```
  switch (lock_result) {
    case LockTableResult::kAlreadyInOwned: {
      return true;
    }
    case LockTableResult::kPlacedInOwned: {
      const TransactionTableResult transaction_result
          = transaction_table_->putOwnEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInOwned)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.1).";
      return true;
    }
    case LockTableResult::kPlacedInPending: {
      const TransactionTableResult transaction_result =
        transaction_table_->putPendingEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInPending)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.2).";
      return false;
    }
    case LockTableResult::kAlreadyInPending:  // Fall through.
    default:
      return false;
  }
```

An alternative tricky implementation would be:

```
  switch (lock_result) {
    case LockTableResult::kPlacedInOwned: {
      const TransactionTableResult transaction_result
          = transaction_table_->putOwnEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInOwned)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.1).";
      // Fall through.
    }
    case LockTableResult::kAlreadyInOwned:
      return true;
    case LockTableResult::kPlacedInPending: {
      const TransactionTableResult transaction_result =
        transaction_table_->putPendingEntry(tid, rid, access_mode);
      CHECK(transaction_result == TransactionTableResult::kPlacedInPending)
          << "Unexpected result in LockManager.acquireLockInternal: "
             "Mismatch of table results (No.2).";
      // Fall through.
    }
    case LockTableResult::kAlreadyInPending:  // Fall through.
    default:
      return false;
  }
```